### ISSUE TITLE
Jacobian uses a lot of memory

### ISSUE THREAD

**Issue by sjperkins** [google/jax#215]
The following script uses around 8GB of memory on my laptop, but the output shape is 6.5MB at most.

```python
import numpy as onp
import jax.numpy as np
from jax import jacrev, jit

R = onp.random.RandomState(0).randn

lm = R(10, 2)
uvw = R(128, 3)
freq = np.linspace(.856e9, .856e9*2, 16)


def fn(lm, uvw, frequency):
    l = lm[:, 0, None, None]
    m = lm[:, 1, None, None]
    n = np.sqrt(1.0 - l**2 - m**2) - 1.0

    u = uvw[None, :, 0, None]
    v = uvw[None, :, 1, None]
    w = uvw[None, :, 2, None]

    return (u*l + v*m + w*n)*frequency


jacobian = jit(jacrev(fn, 0))
print(fn(lm, uvw, freq).shape)
print(np.asarray(jacobian(lm, uvw, freq)).shape)
```

**Comment by mattjj** [comment]
You should probably use `jacfwd` here.

The output is shape `(10, 128, 16)` (total dimension 20480) and the input is shape `(10, 2)` (total dimension 20), so the Jacobian in question has output dimension 20480 and input dimension 20. In general, forward-mode is more FLOP efficient than reverse-mode when the number of output dimensions exceeds the number of input dimensions (i.e. the Jacobian is "tall"). The Jacobian here is effectively of shape `(20480, 20)`, so that's pretty tall!

In this case, the big win is not having to construct a basis for the output space to pull back (i.e. calling `onp.eye(20480)`, as `jacrev` does) and instead just having to construct an input basis to push forward (i.e. calling `onp.eye(20)`, as `jacfwd` does). But in general forward-mode has another memory advantage: since forward-mode is evaluated in the same order as the original function, it only has 2x the maximum memory cost of the original function. That's in contrast with reverse-mode, which has memory cost that scales with the *depth* of the original function (i.e. reverse-mode has to save the linearization points and reuse them on the way back). (There are hybrid strategies one can deploy, e.g. [to get forward-mode's memory savings for elementwise kernels within reverse-mode](https://arxiv.org/abs/1810.08297), but that's when the overall Jacobian to be evaluated is wide enough that you can only do reverse-mode, whereas here forward-mode is better in all respects.)

In my test just now, `jacfwd` was both much faster and used only a tiny amount of memory. Can you try it out and let us know whether `jacfwd` works for you?

**Comment by sjperkins** [comment]
Thanks, jacfwd is much faster!
